### DISCUSSION TITLE
Discord py multiple people using commands at the same time

### DISCUSSION THREAD

**Discussion by 0J33** [Rapptz/discord.py#9057] [Q&A]
I'm fairly new to python and I'm working on a discord bot using discord py that uses the FastF1 api and gets data from it and turns it into plots using matplotlib and seaborn. The problem is that I can't make it so that multiple people can be using the bot at the same time because when a command is being used if you try to use a command it fails to send the message and says "The application did not respond". I've tried to look for solutions using multithreading and running each command in the executor but matplotlib doesn't work all the time when I do that.

Is there any way to make it so that multiple commands can be used while the bot is "thinking" or executing the code for that command? These commands take about a minute but with more users and more load it needs a solution. Can I do it with discord py or do I need to switch do a different language?

Example code for command:

```
@client.tree.command(name="example", description="")
async def example(interaction: discord.Interaction):
  embed = discord.Embed(title="example text", description="example text",colour=discord.Color.dark_blue())
  await interaction.response.defer(ephemeral = False)
  asyncio.sleep(4)

  #get the data from the api and turn it into a graph using matplotlib or seaborn and save it as example.png

  file = discord.File(path/to/file/example.png)
  embed = discord.Embed(title="", description="",colour=discord.Color.dark_blue())
  await interaction.followup.send(file=file, embed=embed, ephemeral = False)
```

**Comment by Rapptz** [comment]
This question is really broad, but fundamentally you need to read up more on asynchronous programming and make sure you're not blocking as defined in [the FAQ](https://discordpy.readthedocs.io/en/latest/faq.html#what-does-blocking-mean). 

You can circumvent blocking by doing a few things, the easiest is to swap your HTTP calls from `requests` or similar to `aiohttp` (or `httpx`'s async client). You can also offload things to another thread using [`asyncio.to_thread`](https://docs.python.org/3/library/asyncio-task.html#asyncio.to_thread) if it involves I/O.

**Reply by 0J33** [reply]
I used `asyncio.to_thread` to offload the blocking parts in the code and it worked. Thanks!

**Comment by felipeyx** [comment]
discord.py can handle multiple commands concurrently. The important part is not blocking the asyncio event loop with CPU-heavy or synchronous work.

First, defer the interaction:

await interaction.response.defer()

Then move the expensive FastF1/matplotlib work away from the Discord event loop:

path = await asyncio.to_thread(generate_plot, arguments)
await interaction.followup.send(file=discord.File(path))

If matplotlib gives problems when used from threads, a ProcessPoolExecutor is a better option for CPU-heavy plotting.

Also this:

asyncio.sleep(4)

does nothing unless it is awaited:

await asyncio.sleep(4)

One more important detail: don't have every invocation write to the same example.png. Give each interaction a unique temporary filename, otherwise two users running the command simultaneously can overwrite each other's graph.
